modules/transport-netty4/src/main/java/org/elasticsearch/transport/SharedGroupFactory.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    public Settings getSettings() {
+        return settings;
+    }
+
+    public int getHttpWorkerCount() {
+        if (httpWorkerCount == 0) {
+            return workerCount;
+        } else {
+            return httpWorkerCount;
+        }
+    }
+
+    public int getTransportWorkerCount() {
+        return workerCount;
+    }
+
+    public synchronized SharedGroup getTransportGroup() {
+        return getGenericGroup();
+    }
+
+    public synchronized SharedGroup getHttpGroup() {
+        if (httpWorkerCount == 0) {
+            return getGenericGroup();
+        } else {
+            NioEventLoopGroup eventLoopGroup = new NioEventLoopGroup(httpWorkerCount,
+                daemonThreadFactory(settings, HttpServerTransport.HTTP_SERVER_WORKER_THREAD_NAME_PREFIX));
+
+            return new SharedGroup(new RefCountedGroup(eventLoopGroup));
+        }
+    }
+
+    private SharedGroup getGenericGroup() {
+        if (refCountedGroup == null) {
+            EventLoopGroup eventLoopGroup = new NioEventLoopGroup(workerCount,
+                daemonThreadFactory(settings, TcpTransport.TRANSPORT_WORKER_THREAD_NAME_PREFIX));
+            this.refCountedGroup = new RefCountedGroup(eventLoopGroup);
+            return new SharedGroup(refCountedGroup);
+        } else {
+            refCountedGroup.incRef();
+            return new SharedGroup(refCountedGroup);
+        }
+    }
+
+    private static class RefCountedGroup extends AbstractRefCounted {
+
+        public static final String NAME = "ref-counted-event-loop-group";
+        private final EventLoopGroup eventLoopGroup;
+
+        private RefCountedGroup(EventLoopGroup eventLoopGroup) {
+            super(NAME);
+            this.eventLoopGroup = eventLoopGroup;
+        }
+
+        @Override
+        protected void closeInternal() {
+            eventLoopGroup.shutdownGracefully(0, 5, TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Wraps the {@link RefCountedGroup}. Calls {@link RefCountedGroup#decRef()} on close. After close,
+     * this wrapped instance can no longer be used.
+     */
+    public static class SharedGroup {
+
+        private final RefCountedGroup refCountedGroup;
+
+        private final AtomicBoolean isOpen = new AtomicBoolean(true);
+
+        private SharedGroup(RefCountedGroup refCountedGroup) {
+            this.refCountedGroup = refCountedGroup;
+        }
+
+        public EventLoopGroup getLowLevelGroup() {
+            return refCountedGroup.eventLoopGroup;
+        }
+
+        public Future<?> shutdownGracefully() {
+            if (isOpen.compareAndSet(true, false)) {
+                refCountedGroup.decRef();
+                if (refCountedGroup.refCount() == 0) {
+                    refCountedGroup.eventLoopGroup.terminationFuture();
+                    return refCountedGroup.eventLoopGroup.terminationFuture();
+                } else {
+                    return getSuccessPromise();
```

Comment:
this is a bit odd. It means that the group is not truly closed and we still report success. I think we should always return the terminationFuture. We can then also get rid of this class.

Reply:
This comment and the lower comment are kind of commenting on the lifecycle of this whole this.

If the transport closes a low-level Netty EventLoopGroup (opposed to using reference counting) it will also close the event loop for the http transport which seems unideal. Closing the SharedGroup ensures that the event loops do not stop until all shared instances are done with them.

Reply:
I was not advocating to get rid of the refcounting, but just getting rid of the SharedGroup class. You have already implemented it in a way that you're not blocking on the future returned by `shutdownGracefully` in the `Netty4HttpTransport`. Might as well return the terrminationFuture in that case and only do the await call in internal transport, relying on the ordered shutdown of the transports.